Back to the Maguture: Reverting to Magees rampup, but cleaned up! (#19)
RIT-EVT/rampup: docs/GLOSSARY.md

Reviewer: The anchor links in this document need to be re-assessed. Markdown processors handle header links with spaces and special characters by URL encoding them. Many of the links to headers in this doc send you nowhere.

Author: i am stupid and cant reproduce so i am ignoring it

happen. Writing drivers for the TMP117 and MAX22530 is the core of Phases 2
and 3.

### Flashing

The process of writing your compiled program onto a microcontroller's
memory so it actually runs there.

### .bin file

Short for binary file. The actual [compiled](#compiler) output you flash
onto a board: raw machine code and nothing else, no symbol names or debug
info attached. This is the file `docs/GETTING_STARTED.md` has you drag
onto a Nucleo or select in STM32CubeProgrammer.

### OpenOCD

OCD stands for On-Chip Debugging. OpenOCD is an open-source OCD tool that 
connects to a board over SWD/JTAG (through the [ST-Link](#st-link)) and lets
you debug your code while it's actually running on the microcontroller: 
pause execution, walk through the program line by line, and inspect variable values in real time. 
Its extremely helpful when trying to debug a problem in your code.

## Working with Data

### Bit

The smallest piece of data; either `0` or `1`.

### Byte

Eight [bits](#bit). A byte can represent 256 different values (`0` to `255`
[unsigned](#unsigned-vs-signed), or `-128` to `127` [signed](#unsigned-vs-signed)).

### Unsigned vs Signed

> *Unsigned* means the data can only be positive (`0` to *x*). *Signed* means

Author: ```suggestion
*Unsigned* means the data can only be positive, whereas *signed* means
```

> it can be negative or positive (`-y` to `y - 1`, where `y = x / 2`), since

Reviewer: > (`-y` to `y - 1`, where `y = x / 2`)

What is this math trying to show? I think its that you can have a maximum value of half that of the unsigned version? If so I think it may be better to say it in plaintext

> it can be negative or positive (`-y` to `y - 1`, where `y = x / 2`), since

Reviewer: > (`-y` to `y - 1`, where `y = x / 2`)

I think this is trying to explain that signed values can only represent half of what an unsigned value can. But it took a while to get, maybe this can just be plaintext instead.

> it can be negative or positive (`-y` to `y - 1`, where `y = x / 2`), since

Author: ```suggestion
it can be negative or positive (i.e. in 4 bits, while signed you can do `[-8, 7]`, when unsigned, you can do `[0, 15]`), since
```

[bit](#bit) with the highest value (the one farthest to the left). It
contributes the most to the number's value, the same way the leftmost
digit does in a decimal number (in 523, that's the `5`).

### LSB

Short for Least Significant Bit. The [bit](#bit) with the least value (the
one farthest to the right), contributing the least to the number's value,
the same way the rightmost digit does in a decimal number (in 523, that's
the `3`).

## Communication Fundamentals

### Bus

A set of wires that lets different parts of a system communicate, shared by
more than two devices (as opposed to a simple point-to-point wire between
just two). The CAN bus is what lets every board on the bike talk over one
shared set of wires.

### Serial Communication

Data is sent one [bit](#bit) at a time, over a single wire (or pair of
wires). [UART](#uart), [I2C](#i2c), and [SPI](#spi) are all serial.

### Parallel Communication

Many bits of data are sent at the same time, each over its own wire.

### Half-Duplex

Communication where data can flow in both directions, but only one
direction at a time, like a walkie-talkie. [I2C](#i2c)'s data line (SDA) is

> shared this way between master and slave.

Reviewer: ```suggestion
shared this way between leader and follower.
```

### Full-Duplex

Communication where data can flow in both directions at the same time,
like a phone call, because each direction gets its own dedicated wire.
[UART](#uart) (separate TX/RX lines) and [SPI](#spi) (separate MOSI/MISO
lines) are both full-duplex.

### Baud Rate

The speed of data transmission, usually measured in bits per second (bps),
for [serial communication](#serial-communication) like [UART](#uart).
Rampup's UART phase uses 9600 baud.

### Communication Protocol

A set of rules two devices agree on ahead of time so they can reliably
exchange data: things like who talks first, how fast, and how bits are
grouped into meaningful values.

### Master / Slave

A communication model where the master decides when communication happens
and who it talks to. Slave devices wait and respond only when the master

> addresses them. [I2C](#i2c) and [SPI](#spi) both use this model.

Reviewer: ```suggestion
### Leader (Master) / Follower (Slave)

A communication model where the leader decides when communication happens
and who it talks to. Follower devices wait and respond only when the leader
addresses them. [I2C](#i2c) and [SPI](#spi) both use this model.
```

Just updating to the newer conventions of not using master / slave

> addresses them. [I2C](#i2c) and [SPI](#spi) both use this model.

Reviewer: ```suggestion
### Leader (Master) / Follower (Slave)

A communication model where the leader decides when communication happens
and who it talks to. Follower devices wait and respond only when the leader
addresses them. [I2C](#i2c) and [SPI](#spi) both use this model.
```

We should be moving towards the modern convention of Leader / Follower over Master / Slave

An *oscillator* is the physical hardware component that generates that
signal.

### Analog vs Digital

An *analog* signal is continuous, it can take any value in a range (like
the voltage out of a sensor). A *digital* signal is discrete, it only takes
specific values, most commonly just two (`1`/`0`, "high"/"low"). An ADC
(see Phase 3) converts an analog signal into a digital one your
microcontroller can work with.

### PWM

Short for Pulse Width Modulation. A digital signal rapidly switched on and
off, where the fraction of time it's "on" (the duty cycle) sets its average
value. Used to fake an analog output (control brightness, speed, etc.)
using only a [GPIO](#gpio) pin.

## Protocols You'll Use in Rampup

### UART

*Phase 1*

Short for Universal Asynchronous Receiver/Transmitter. A simple,
asynchronous [serial](#serial-communication)
[protocol](#communication-protocol) with no shared clock signal; both sides
just have to agree on the [baud rate](#baud-rate) ahead of time. Used
mainly for laptop-to-board debugging output.

### I2C

*Phase 2*

Short for Inter-Integrated Circuit, read "eye-two-cee" or "eye-squared-cee."

> A two-wire, synchronous, [master/slave](#master-slave)

Reviewer: ```suggestion
A two-wire, synchronous, [leader/follower](#Leader%20%28Master%29%20%2F%20Follower%20%28Slave%29)
```

Markdown links need to be url encoded to actually pick up rich titles

[serial](#serial-communication) protocol: one wire for data (SDA) and one
for a shared clock (SCL), which tells the devices when to read the value from the data line. 

> Multiple slave devices can share the same two wires, each addressed individually. 

Reviewer: ```suggestion
Multiple follower devices can share the same two wires, each addressed individually. 
```

Used in rampup to read the TMP117 temperature sensor.

### SPI

*Phase 3*

Short for Serial Peripheral Interface, read "spy." A faster, synchronous,

> [master/slave](#master-slave) [serial](#serial-communication) protocol

Reviewer: ```suggestion
[leader/follower](#Leader%20%28Master%29%20%2F%20Follower%20%28Slave%29) [serial](#serial-communication) protocol
```

using separate data lines for each direction (MOSI/MISO), a shared clock

> (SCK), and a dedicated chip-select (CS) line per slave device. Used in

Reviewer: ```suggestion
(SCK), and a dedicated chip-select (CS) line per follower device. Used in
```

rampup to talk to the MAX22530 ADC.

### CAN

*Phase 4*

Short for Controller Area Network. A serial bus, originally built for cars,
where every device on the bus sees every message and decides for itself
whether to care about it (unlike [I2C](#i2c)/[SPI](#spi), there's no single

> fixed master). It's the shared bus EVT's boards use to talk to each other

Reviewer: ```suggestion
fixed leader). It's the shared bus EVT's boards use to talk to each other
```

| [CAN](#can) | Controller Area Network                                                                                                                                                               |
| CLK | Clock, a generic name for a clock signal line. See [Clock / Oscillator](#clock--oscillator).                                                                                          |
| [COB-ID](#cob-id) | CAN message ID used for a specific purpose                                                                                                                                            |
| CPHA | Clock Phase, part of an SPI device's mode setting (paired with CPOL). A fiddly timing detail you'll rarely have to figure out yourself, the datasheet will tell you the right value.  |
| CPOL | Clock Polarity, part of an SPI device's mode setting (paired with CPHA). Same idea as CPHA, another timing detail the datasheet spells out for you.                                   |
| CRC | Cyclic Redundancy Check: error-checking bits included at the end of a CAN frame, like a checksum that catches data that got corrupted along the way.                                  |

> | CS | Chip Select: the SPI line that tells one follower device it's being addressed, like tapping one specific person on the shoulder before talking to them.                               |

Reviewer 2: This one is phrased a little weirdly

| DBC | Database CAN: a file format for documenting what a CAN message's bytes mean.                                                                                                          |
| [FTB](#ftb) | Firmware Test Board                                                                                                                                                                   |
| [GND](#gnd) | Ground                                                                                                                                                                                |
| [GPIO](#gpio) | General Purpose Input/Output                                                                                                                                                          |
| HAL | Hardware Abstraction Layer: ST's own driver layer that EVT-core is built on top of, so you get to call simple functions instead of poking at raw hardware registers yourself.         |
| [HUDL](#hudl) | Phase 4's display node                                                                                                                                                                |

> | [I2C](#i2c) | Inter-Integrated Circuit                                                                                                                                                              |

Reviewer 2: Inter-Integrated Circuit. A Common protocol used to interface with devices

Reviewer 2: oh its got a link to it that tells you what it has
